**Evictions Dataset Description**

This dataset lists pending, scheduled, and executed evictions within the five boroughs,
for the year 2017 - early 2022. The data fields may be sorted by Court Index Number,
Docket Number, Eviction Address, Apartment Number, Executed Date, Marshal First
Name, Marshal Last Name, Residential or Commercial (property type), Borough, Zip
Code and Scheduled Status (Pending/Scheduled).

1. Import pandas  

In [ ]:
import pandas as pd

2. Read CSV into DataFrame

In [ ]:
evictions_df = pd.read_csv('Evictions.csv')
evictions_df

Court Index Number  Docket Number          Eviction Address  \
0               50365/19          352435     319 WEST 94TH STREET   
1             B068159/16          379048         2332 CRESTON AVE   
2               59891/16          320691   2670 BAINBRIDGE AVENUE   
3               75708/18          115775          18-24 25TH ROAD   
4              900940/18           86395  3005 EASTCHESTER  RO AD   
...                  ...             ...                      ...   
67723         R 50452/17           68450     182 ARLINGTON AVENUE   
67724          N66489/17           81749   600 WEST 136TH STREE T   
67725         60675/17-1          208636         115 LINCOLN ROAD   
67726          B42335/17           83318  1131-1133 OGDEN AVEN UE   
67727           20290/16           59510             5 FLEET WALK   

      Eviction Apartment Number Executed Date Marshal First Name  \
0                          C103    03/25/2019             Thomas   
1                            41    04/25/2017            Richard   
2                            2F    02/21/2017               John   
3                             1    04/23/2019             Maxine   
4                    STOREFRONT    11/08/2018             Justin   
...                         ...           ...                ...   
67723                       NaN    03/29/2017             Steven   
67724                        6F    09/27/2017             Ileana   
67725                        3N    01/09/2020            Richard   
67726                       12A    12/13/2017             Ileana   
67727                       08E    08/02/2017             Justin   

      Marshal Last Name Residential/Commercial        BOROUGH  \
0                   Bia            Residential      MANHATTAN   
1                 McCoy            Residential          BRONX   
2            Villanueva            Residential          BRONX   
3              Chevlowe            Residential         QUEENS   
4              Grossman             Commercial          BRONX   
...                 ...                    ...            ...   
67723            Powell            Residential  STATEN ISLAND   
67724            Rivera            Residential      MANHATTAN   
67725           Capuano            Residential       BROOKLYN   
67726            Rivera            Residential          BRONX   
67727          Grossman            Residential       BROOKLYN   

       Eviction Postcode         Ejectment Eviction/Legal Possession  \
0                  10025  Not an Ejectment                Possession   
1                  10468  Not an Ejectment                Possession   
2                  10458  Not an Ejectment                Possession   
3                  11102  Not an Ejectment                Possession   
4                  10469  Not an Ejectment                Possession   
...                  ...               ...                       ...   
67723              10303  Not an Ejectment                Possession   
67724              10031  Not an Ejectment                Possession   
67725              11225  Not an Ejectment                Possession   
67726              10452  Not an Ejectment                Possession   
67727              11201  Not an Ejectment                Possession   

        Latitude  Longitude  Community Board  Council District  Census Tract  \
0      40.794205 -73.974734              7.0               6.0         183.0   
1      40.858643 -73.900402              5.0              14.0       23703.0   
2      40.865470 -73.891472              7.0              15.0       40502.0   
3      40.774861 -73.926140              1.0              22.0          91.0   
4      40.869607 -73.842766             12.0              12.0         358.0   
...          ...        ...              ...               ...           ...   
67723  40.635485 -74.167531              1.0              49.0         323.0   
67724  40.820935 -73.954996              9.0               7.0       22301

3. Explore the data and its structure

In [ ]:
# .shape shows the number of rows and columns
evictions_df.shape

(67728, 20)

In [ ]:
# .dtypes shows the data types
evictions_df.dtypes

Court Index Number            object
Docket Number                  int64
Eviction Address              object
Eviction Apartment Number     object
Executed Date                 object
Marshal First Name            object
Marshal Last Name             object
Residential/Commercial        object
BOROUGH                       object
Eviction Postcode              int64
Ejectment                     object
Eviction/Legal Possession     object
Latitude                     float64
Longitude                    float64
Community Board              float64
Council District             float64
Census Tract                 float64
BIN                          float64
BBL                          float64
NTA                           object
dtype: object

In [ ]:
# .isnull() detects missing values
# To get the total summation of all missing values in the dataframe,
# chain two .sum() methods together
evictions_df.isnull().sum().sum()

62997

The DataFrame contains 62997 missing values. We want to further investigate where these missing values are.  

In [ ]:
# .isna() detects missing values in columns
evictions_df.isna().sum()

Court Index Number               0
Docket Number                    0
Eviction Address                 0
Eviction Apartment Number    11209
Executed Date                    0
Marshal First Name               0
Marshal Last Name                0
Residential/Commercial           0
BOROUGH                          0
Eviction Postcode                0
Ejectment                        0
Eviction/Legal Possession        0
Latitude                      6444
Longitude                     6444
Community Board               6444
Council District              6444
Census Tract                  6444
BIN                           6562
BBL                           6562
NTA                           6444
dtype: int64

Missing values:


*   11209 missing values in the 'Eviction Apartment Number' column
*   6444 missing values in the 'Latitude' column
* 6444 missing values in the 'Longitude' column
* 6444 missing values in the 'Community Board' column
* 6444 missing values in the 'Council District' column
* 6444 missing values in the 'Census Tract' column
* 6562 missing values in the 'BIN' column
* 6562 missing values in the 'BBL' column
* 6444 missing values in the 'NTA' column



4. Investigate the missing values and clean the data

In [ ]:
# Missing values in the Eviction Apartment Number column
# Not all Eviction Addresses are apartments
# Hence we will ignore this

In [ ]:
# Looking at the number of missing values across all columns it is
# possible that the rows with the missing values in 'Latitude' are also the
# same rows with missing values in 'Longitude', 'Community Board', 'Council District',
# 'Census Tract', 'BIN', 'BBL', and 'NTA'

# Hence we are going to drop the rows with null values in the 'Latitude' column

# .dropna() removes all rows that contain NULL values
# subset drops rows with NA values in a specific column
# reset_index(drop=True) will re-index the index after sorting
evictions_df = evictions_df.dropna(subset=['Latitude']).reset_index(drop=True)

In [ ]:
# View the changes in the number of missing values across all columns after dropping rows with
# null values in the 'Latitude' column

# .isna() detects missing values in columns
evictions_df.isna().sum()

Court Index Number              0
Docket Number                   0
Eviction Address                0
Eviction Apartment Number    7364
Executed Date                   0
Marshal First Name              0
Marshal Last Name               0
Residential/Commercial          0
BOROUGH                         0
Eviction Postcode               0
Ejectment                       0
Eviction/Legal Possession       0
Latitude                        0
Longitude                       0
Community Board                 0
Council District                0
Census Tract                    0
BIN                           118
BBL                           118
NTA                             0
dtype: int64

By dropping the rows with null values in the 'Latitude' column, we noticed a decrease in the number of missing values in the 'Eviction Apartment Number' column as well as the 'BIN' and 'BBL' columns. Furthermore, the number of missing values for the columns 'Longitude', 'Community Board', 'Council District', 'Census Tract', and 'NTA' are now 0.


In [ ]:
# View the change in the dimension of the DataFrame
# .shape shows the number of rows and columns
evictions_df.shape

(61284, 20)

In [ ]:
# As mentioned before, we will ignore the missing values in
# 'Eviction Apartment Number'

# Looking at the BIN and BBL columns, there are 118 missing values for both


# BIN - The Building Identification Number
# BBL - The Borough, Block, Lot

# The values for the columns above are unique to its location
# We do not possess information to fill in the missing values

# We are going to drop the rows with null values in the 'BIN' column
evictions_df = evictions_df.dropna(subset=['BIN']).reset_index(drop=True)

In [ ]:
# View the changes in the number of missing values across all columns after dropping rows with
# null values in the 'BIN' column

# .isna() detects missing values in columns
evictions_df.isna().sum()

Court Index Number              0
Docket Number                   0
Eviction Address                0
Eviction Apartment Number    7321
Executed Date                   0
Marshal First Name              0
Marshal Last Name               0
Residential/Commercial          0
BOROUGH                         0
Eviction Postcode               0
Ejectment                       0
Eviction/Legal Possession       0
Latitude                        0
Longitude                       0
Community Board                 0
Council District                0
Census Tract                    0
BIN                             0
BBL                             0
NTA                             0
dtype: int64

By dropping the rows with null values in the 'BIN' column, we noticed that there is now no missing values in the 'BBL' column.

This is most likely because the rows with missing BIN values are also missing BBL values.

In [ ]:
# View the change in the dimension of the DataFrame
# .shape shows the number of rows and columns
evictions_df.shape

(61166, 20)

We've handled most of the nulls and cleaned our data. It is time to remove any duplicate rows.

In [ ]:
# drop_duplicates() removes duplicate rows based on all columns
evictions_df = evictions_df.drop_duplicates()

In [ ]:
# View the change in the dimension of the DataFrame
# .shape shows the number of rows and columns
evictions_df.shape

(60645, 20)

5. Now that we have cleaned the data, let's write the cleaned table to a csv file.

In [ ]:
evictions_df.to_csv("evictions_clean.csv", index=False)